addons/udes_stock_routing/models/stock_picking.py

```diff
@@ -0,0 +1,174 @@
+                    # Set the fallbacks to optional configurations here so grouping is simpler.
+                    intermediate_dest_location_id=config_location.u_two_stage_intermediate_dest_location.id
+                    or self.location_dest_id.id,
+                    operation_type_id=config_location.u_two_stage_intermediate_operation_type.id
+                    or self.picking_type_id.id,
+                )
+            ] |= move_line.move_id
+
+        return configuration_move_mapping
+
+    def initiate_two_stage_split(self):
+        """
+        Perform all setup to turn a one stage outbound picking to potentially multiple two stage pickings,
+        as per configuration of the reserved moves locations.
+        """
+        self.ensure_one()
+        Picking = self.env["stock.picking"]
+
+        original_operation_type = self.picking_type_id
+        original_location = self.location_id
+        original_priority = self.priority
+        configuration_move_mapping = self._get_two_stage_configuration_move_mapping()
+        # Unreserve stock on the current picking before proceeding.
+        self.do_unreserve()
+        for configuration, moves in configuration_move_mapping.items():
+            stage_1_pick = Picking.create(
+                {
+                    "picking_type_id": configuration.operation_type_id,
+                    "location_id": original_location.id,
+                    "location_dest_id": configuration.intermediate_location_id,
+                    "scheduled_date": fields.Datetime.now(),
+                    "priority": original_priority,
+                    "u_from_two_stage_split": True,
+                }
+            )
+            stage_2_pick = Picking.create(
+                {
+                    "picking_type_id": original_operation_type.id,
+                    "location_id": configuration.intermediate_location_id,
+                    "location_dest_id": configuration.intermediate_dest_location_id,
+                    "scheduled_date": fields.Datetime.now(),
+                    "priority": original_priority,
+                    "u_from_two_stage_split": True,
+                }
+            )
+            for move in moves:
+                # Duplicate the move to both stages, updating the source, dest, and move orig/dest links accordingly.
+                # move_dest_ids gets cleared on copy()
+                stage_1_move = move.copy(
+                    default={
+                        "location_dest_id": configuration.intermediate_location_id,
+                        "picking_id": stage_1_pick.id,
+                        "picking_type_id": configuration.operation_type_id,
+                        # preserve move_orig_ids, incase this move is not first-in-chain, as the original move will be removed.
+                        "move_orig_ids": [[6, 0, move.move_orig_ids.ids]],
+                    }
+                )
+                _stage_2_move = move.copy(
```

Comment:
I will change this to modify the original move with the updated links / locations instead of copying it, to avoid needing to cancel the original moves at the end which results in additional cancellations further on in the pick chain. However, the change introduces more breakages in unit tests so I will fix those when I get back